library/manageiq_tag_assignment.py

```diff
@@ -0,0 +1,225 @@
+        """
+        entities_list = getattr(self.client.collections, entity_type)
+        return next((e.id for e in entities_list if e.name == entity_name), None)
+
+    def query_resource_tags(self, resource_type, resource_id):
+        """ Returns the tags assigned to the resource.
+        """
+        try:
+            url = '{api_url}/{resource_type}/{resource_id}/tags?expand=resources'.format(api_url=self.api_url, resource_type=resource_type, resource_id=resource_id)
+            result = self.client.get(url)
+            return result.get('resources', [])
+        except Exception as e:
+            self.module.fail_json(msg="Failed to query {resource_type} tags: {error}".format(resource_type=resource_type, error=e))
+
+    def is_tag_assigned(self, resource_type, resource_id, category_name, tag_name):
+        """Return True if the tag is assigned on the resource, False otherwise.
+        """
+        assigned_tags = self.query_resource_tags(resource_type, resource_id)
+        full_tag_name = '/managed/{category_name}/{tag_name}'.format(category_name=category_name, tag_name=tag_name)
+        return any(tag['name'] == full_tag_name for tag in assigned_tags)
+
+    def execute_action(self, resource_type, resource_id, tags, action):
+        """Executes the action for the resource tag
+        """
+        url = '{api_url}/{resource_type}/{resource_id}/tags'.format(api_url=self.api_url, resource_type=resource_type, resource_id=resource_id)
+        try:
+            result = self.client.post(url, action=action, resources=tags)
+        except Exception as e:
+            self.module.fail_json(msg="Failed to {action} tag: {error}".format(action=action, error=e))
+        for result in result['results']:
+            if result['success']:
+                self.changed = True
+            else:
+                self.module.fail_json(msg="Failed to {action}: {fail_message}".format(action=action, entity=entity, fail_message=result['results'][0]['message']))
+
+    def assign_or_unassign_tag(self, tags, resource, resource_name, state):
+        """ Assign or unassign the tag on a manageiq resource.
+
+        Returns:
+            Whether or not a change took place and a message describing the
+            operation executed.
+        """
+        resource_type = self.manageiq_entities[resource]
+        resource_id = self.find_entity_by_name(resource_type, resource_name)
+        if not resource_id:  # resource doesn't exist
+            self.module.fail_json(
+                msg="Failed to {action} tag: {resource_name} {resource} does not exist in manageiq".format(action=ManageIQ.actions[state], resource_name=resource_name, resource=resource))
+
+        tags_to_execute = []
+        for tag in tags:
+            tag_name = tag['name']
+            category_name = tag['category']
+            assigned = self.is_tag_assigned(resource_type, resource_id, category_name, tag_name)
```

Comment:
This queries resource's all tags, looks for one among them, repeats...
Can you query once, build a set, and do set difference / intersection?